core/fetcher.go

```diff
@@ -0,0 +1,87 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/celestiaorg/celestia-core/types"
+	"github.com/celestiaorg/celestia-node/service/block"
+)
+
+const newBlockSubscriber = "NewBlock/Events"
+
+var newBlockEventQuery = types.QueryForEvent(types.EventNewBlock).String()
+
+type BlockFetcher struct {
+	client Client
+
+	mux        sync.Mutex
+	newBlockCh chan *block.Raw
+}
+
+// NewBlockFetcher returns a new `BlockFetcher`.
+func NewBlockFetcher(client Client) *BlockFetcher {
+	return &BlockFetcher{
+		client: client,
+	}
+}
+
+// GetBlock queries Core for a `Block` at the given height.
+func (f *BlockFetcher) GetBlock(ctx context.Context, height *int64) (*block.Raw, error) {
+	raw, err := f.client.Block(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+	return raw.Block, nil
+}
+
+// SubscribeNewBlockEvent subscribes to new block events from Core, returning
+// a new block event channel on success.
+func (f *BlockFetcher) SubscribeNewBlockEvent(ctx context.Context) (<-chan *block.Raw, error) {
+	// start the client if not started yet
+	if !f.client.IsRunning() {
+		return nil, fmt.Errorf("client not running")
+	}
+	eventChan, err := f.client.Subscribe(ctx, newBlockSubscriber, newBlockEventQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a wrapper channel for translating ResultEvent to "raw" block
+	if f.newBlockCh != nil {
+		return nil, fmt.Errorf("new block event channel exists")
+	}
+
+	f.mux.Lock()
+	f.newBlockCh = make(chan *block.Raw)
+	f.mux.Unlock()
```

Comment:
Why is the channel guarded by a mutex?

Reply:
I see this is coming from here: https://github.com/celestiaorg/celestia-node/pull/59#discussion_r709215476
@Wondertan can you explain why the channel should be guarded by a mutex? I don't see why this makes sense here.

Reply:
@mattdf also mentioned that this is unnecessary during our pairing session. I don't mind either way -- I also don't see a circumstance (at least now) this chan will be written to / accessed at the same time.

Reply:
@liamsi, BlockFetcher is currently used only once in the app - in the block service. In the future, other components may rely on the same BlockFetcher, and chances that we will forget to put the lock are pretty high. That's a safety measure that doesn't hurt. Furthermore, It is not clear how block service will be implemented. It might be accessed from multiple routines where one starts the subscription and the other stops it causing the race on the `newBlockCh`. I've been in such situations before and it's better to avoid such races in advance.

Reply:
> In the future, other components may rely on the same BlockFetcher, and chances that we will forget to put the lock are pretty high.

Yeah, but even then, there won't be multiple go-routines calling subscribe/unsubscribe concurrently. Let's remove the mutex. The go race detector is pretty good and we would notice eventually even if we forgot to add it in the future (which from how I understand the blockfetcher being used won't be necessary anyways).

Reply:
I agree that from a pure correctness point of view it does not hurt. But you could argue the same for arbitrary fields. And besides (the marginal computational cost) there is a cognitive cost. E.g. if I see a mutex on a channel I can't stop but think: wait, why is this guarded by a lock? It's a distraction, implies cognitive overhead, and decreases readability. 

Unless there is a clear access pattern _now_ that justifies that mutex, it should be removed IMO. We can care about the future in the future :)

Reply:
yeah i agree that it should be removed, overreaching "validation" is as wrong as insufficient validation. it's misleading as ismail says, it implies that there will be concurrent reader channels when that should never be the case. if you want to go full pedant there should be an assert here with a panic rather than a mutex.

Reply:
Safety > cognitive overhead!

I think we won't found a middle ground here, so let's just remove it and see how it goes. We will see in the future how it goes.

Reply:
> Safety > cognitive overhead!

I disagree: code with as little cognitive overhead as possible usually is safer than code that is unclear (aka cognitive overhead) IMHO.

Reply:
> if you want to go full pedant there should be an assert here with a panic rather than a mutex.

I disagree that panics are the right solution, but let's not argue and just remove mutext.